File: src/GlobalClientEvent.ts
import { getGlobalRemote as getGlobalRemoteId } from "internal";
import NetClientEvent from "ClientEvent";

/**
 * Client counter-part to GlobalServerEvent
 */
export default class NetGlobalClientEvent {
	private readonly instance: NetClientEvent;
	constructor(name: string) {
		this.instance = new NetClientEvent(getGlobalRemoteId(name));
	}

	/**
	 * Connect a function to fire when the event is invoked by the client
	 * @param callback The function fired when the event is invoked by the client
	 */
	public Connect<T extends Array<any>>(callback: (...args: T) => void) {
		this.instance.Connect(callback as Callback);
	}
}

File: GlobalClientEvent.lua

local TS = _G[script]
local getGlobalRemoteId = TS.import(script, script.Parent, "internal").getGlobalRemote
local NetClientEvent = TS.import(script, script.Parent, "ClientEvent").default
--[[
	*
	* Client counter-part to GlobalServerEvent
]]
local NetGlobalClientEvent
do
	NetGlobalClientEvent = setmetatable({}, {
		__tostring = function()
			return "NetGlobalClientEvent"
		end,
	})
	NetGlobalClientEvent.__index = NetGlobalClientEvent
	function NetGlobalClientEvent.new(...)
		local self = setmetatable({}, NetGlobalClientEvent)
		self:constructor(...)
		return self
	end
	function NetGlobalClientEvent:constructor(name)
		self.instance = NetClientEvent.new(getGlobalRemoteId(name))
	end
	function NetGlobalClientEvent:Connect(callback)
		self.instance:Connect(callback)
	end
end
return {
	default = NetGlobalClientEvent,
}
